sdk/Pulumi.Automation/Commands/LocalPulumiCmd.cs

```diff
@@ -5,18 +5,243 @@
+    /// </summary>
+    public class LocalPulumiCmdOptions
+    {
+        /// <summary>
+        /// The version of the Pulumi CLI to install or the minimum version requirement for an existing installation.
+        /// </summary>
+        public SemVersion? Version { get; set; }
+        /// <summary>
+        /// The directory where to install the Pulumi CLI to or where to find an existing installation.
+        /// </summary>
+        public string? Root { get; set; }
+        /// <summary>
+        /// If true, skips the version validation that checks if an existing Pulumi CLI installation is compatible with the SDK.
+        /// </summary>
+        public bool SkipVersionCheck { get; set; }
+    }
+
+    /// <summary>
+    /// A <see cref="IPulumiCmd"/> implementation that uses a locally installed Pulumi CLI.
+    /// </summary>
+    public class LocalPulumiCmd : IPulumiCmd
     {
+        // TODO: move to shared place with LocalWorkspace
+        private static string SkipVersionCheckVar = "PULUMI_AUTOMATION_API_SKIP_VERSION_CHECK";
+        private static readonly SemVersion _minimumVersion = new SemVersion(3, 1, 0);
+        internal Task ReadyTask { get; }
+        private readonly string _command;
+        private SemVersion? _version;
+
+        /// <summary>
+        /// The version of the Pulumi CLI that is being used.
+        /// </summary>
+        public SemVersion? Version { get => _version; }
+
+        /// <summary>
+        /// Creates a new LocalPulumiCmd instance.
+        /// </summary>
+        /// <param name="options">Options to configure the LocalPulumiCmd.</param>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns></returns>
+        public static async Task<LocalPulumiCmd> CreateAsync(
+            LocalPulumiCmdOptions? options = null,
+            CancellationToken cancellationToken = default)
+        {
+            var cmd = new LocalPulumiCmd(options, cancellationToken);
+            await cmd.ReadyTask.ConfigureAwait(false);
+            return cmd;
+        }
+
+        private LocalPulumiCmd(LocalPulumiCmdOptions? options, CancellationToken cancellationToken)
+        {
+            var readyTasks = new List<Task>();
+
+            if (options?.Root != null)
+            {
+                _command = Path.Combine(options.Root, "bin", "pulumi");
+            }
+            else
+            {
+                _command = "pulumi";
+            }
+
+            var minimumVersion = _minimumVersion;
+            if (options?.Version != null && options.Version > minimumVersion)
+            {
+                minimumVersion = options.Version;
+            }
+
+            var optOut = options?.SkipVersionCheck ?? Environment.GetEnvironmentVariable(SkipVersionCheckVar) != null;
+            readyTasks.Add(SetPulumiVersionAsync(minimumVersion, _command, optOut, cancellationToken));
+            ReadyTask = Task.WhenAll(readyTasks);
```

Comment:
This is a bit of an ick. I'd probably leave the constructor for just field setting and have all the install logic in `CreateAsync`. Then you can just await in the function and there's no need for this odd ReadyTask field.

Reply:
https://github.com/pulumi/pulumi-dotnet/pull/226/commits/c2b66fa013f9068f4d958ac05b5a94914ea983cd